aten/src/ATen/native/cuda/Normalization.cu

```diff
@@ -0,0 +1,444 @@
+
+  int plane = blockIdx.y * blockDim.y + threadIdx.y;
+
+  if (plane >= input.size(1)) {
+    return;
+  }
+
+  accscalar_t gamma = weight.size(0) > 0 ? static_cast<accscalar_t>(weight[plane]) : static_cast<accscalar_t>(1);
+  accscalar_t beta = bias.size(0) > 0 ? static_cast<accscalar_t>(bias[plane]) : static_cast<accscalar_t>(0);
+  accscalar_t mean = static_cast<accscalar_t>(mean_[plane]);
+  accscalar_t invstd;
+  if (train) {
+    invstd = var_or_invstd[plane];
+  } else {
+    invstd = static_cast<accscalar_t>(1) / std::sqrt(static_cast<accscalar_t>(var_or_invstd[plane]) + epsilon);
+  }
+  for (int64_t batch = blockIdx.x; batch < input.size(0); batch += gridDim.x) {
+    for (int64_t feature = blockIdx.z; feature < input.size(2); feature += gridDim.z) {
+      output[batch][plane][feature] = static_cast<scalar_t>(gamma * (input[batch][plane][feature] - mean) * invstd + beta);
+    }
+  }
+}
+
+
+template <typename scalar_t, typename accscalar_t>
+__global__ void batch_norm_collect_statistics_kernel(
+    const PackedTensorAccessor<scalar_t, 3, at::RestrictPtrTraits> input,
+    PackedTensorAccessor<scalar_t, 3, at::RestrictPtrTraits> output,
+    const PackedTensorAccessor<scalar_t, 1, at::RestrictPtrTraits> weight,
+    const PackedTensorAccessor<scalar_t, 1, at::RestrictPtrTraits> bias,
+    const accscalar_t epsilon,
+    const accscalar_t momentum,
+    PackedTensorAccessor<scalar_t, 1, at::RestrictPtrTraits> running_mean,
+    PackedTensorAccessor<scalar_t, 1, at::RestrictPtrTraits> running_var,
+    PackedTensorAccessor<accscalar_t, 1, at::RestrictPtrTraits> save_mean,
+    PackedTensorAccessor<accscalar_t, 1, at::RestrictPtrTraits> save_invstd) {
+
+  int plane = blockIdx.x;
+  int N = input.size(0) * input.size(2);
+
+  accscalar_t norm = accscalar_t(1) / N;
+
+  // Compute the mean and variance across (batch, x/y/z)
+  accscalar_t mean = reduce<accscalar_t>(SumOp<scalar_t, accscalar_t>(input), input, plane) * norm;
+  __syncthreads();
+  accscalar_t varN = reduce<accscalar_t>(VarOp<scalar_t, accscalar_t>(mean, input), input, plane);
+
+  // Save the mean, variance, and moving averages
+  if (threadIdx.x == 0) {
+    accscalar_t invstd = 0;
+    if (varN != static_cast<accscalar_t>(0) || epsilon != static_cast<accscalar_t>(0)) {
+      invstd = static_cast<accscalar_t>(1) / std::sqrt(varN * norm + epsilon);
+    }
+    save_mean[plane] = mean;
+    save_invstd[plane] = invstd;
+    if (running_mean.data() != NULL) {
+      running_mean[plane] = static_cast<scalar_t>((1 - momentum) * running_mean[plane] + momentum * mean);
+    }
+    if (running_var.data() != NULL) {
+      accscalar_t unbiasedVar = varN / (N - 1);
+      running_var[plane] = static_cast<scalar_t>((1 - momentum) * running_var[plane] + momentum * unbiasedVar);
+    }
+  }
+
+}
+
+template <typename scalar_t, typename accscalar_t>
+__global__ void batch_norm_backward_kernel(
+    const PackedTensorAccessor<scalar_t, 3, at::RestrictPtrTraits> input,
```

Comment:
Well I'm actually not sure if these should be `__restrict__` (other than `save_mean` and `save_invstd` and the output tensors). The other ones can be very well provided by the users.

Reply:
My take on this is that the casual users will only provide input themselves, but I'm happy to remove them if you prefer.

Reply:
Let's remove these traits. We shouldn't hurt out power users. :)

Reply:
I don't understand. What's the argument for **not** making them `__restrict__`? Weren't they previously `__restrict__`?

Reply:
I don't think they were `__restrict__` in the old code.
Personally, I don't see why people should set up overlapping tensors to trick themselves, but I can see how it's a bit safer when not assuming they don't.